linera-core/src/client/mod.rs

```diff
@@ -2320,58 +2318,48 @@ impl<Env: Environment> ChainClient<Env> {
 
         let (_, committee) = self.admin_committee().await?;
         let nodes = self.client.make_nodes(&committee)?;
-        let (max_epoch, committees) = self.client.admin_committees().await?;
 
         // Download certificates for each sender chain at the specific heights.
-        let certificates = stream::iter(missing_blocks.into_iter())
+        stream::iter(missing_blocks.into_iter())
             .map(|(sender_chain_id, heights)| {
+                let height = heights.into_iter().max();
                 let nodes = nodes.clone();
+                let this = self.clone();
                 async move {
+                    let Some(height) = height else {
+                        return Ok(());
+                    };
                     // Try to download from any node.
                     for node in &nodes {
-                        if let Ok(certs) = node
-                            .download_certificates_by_heights(sender_chain_id, heights.clone())
+                        if let Err(err) = this
+                            .download_sender_block_with_sending_ancestors(
+                                sender_chain_id,
+                                height,
+                                node,
+                            )
```

Comment:
I think it might create duplicate requests 🤔 If we're missing blocks 3,4,5 and 5 references 4, and 4 references 3 then
- request for 5th one will download 5,4,3
- for the 4th one will download 4,3
- etc